File: utils/flameColor.ts
const STOPS = [0, 120, 240, 360, 480, 600];
const COLORS = ['#FFC107', '#FF9800', '#F44336', '#3B82F6', '#9C27B0', '#4CAF50'];

function hexToRgb(hex: string): [number, number, number] {
  return [
    parseInt(hex.slice(1, 3), 16),
    parseInt(hex.slice(3, 5), 16),
    parseInt(hex.slice(5, 7), 16),
  ];
}

function rgbToHex(r: number, g: number, b: number): string {
  return '#' + [r, g, b]
    .map(v => Math.round(v).toString(16).padStart(2, '0'))
    .join('');
}

export function flameColorForBurn(calories: number): string {
  const clamped = Math.min(Math.max(calories, 0), 600);
  for (let i = 0; i < STOPS.length - 1; i++) {
    if (clamped >= STOPS[i] && clamped <= STOPS[i + 1]) {
      const t = (clamped - STOPS[i]) / (STOPS[i + 1] - STOPS[i]);
      const [r1, g1, b1] = hexToRgb(COLORS[i]);
      const [r2, g2, b2] = hexToRgb(COLORS[i + 1]);
      return rgbToHex(r1 + t * (r2 - r1), g1 + t * (g2 - g1), b1 + t * (b2 - b1));
    }
  }
  return COLORS[COLORS.length - 1];
}

export function glowIntensityForBurn(calories: number): number {
  return Math.min(Math.max(calories, 0), 600) / 600;
}

File: swift/HealthTracker/Design/FlameColor.swift
import SwiftUI

/// Verbatim port of `expo/utils/flameColor.ts` — the activity-burn flame gradient (0...600 cal).
/// 6-stop, channel-wise linear RGB interpolation. Used exclusively by the CalorieFlame view.
enum FlameColor {
    private static let stops: [Double] = [0, 120, 240, 300, 420, 540]
    private static let hexes: [String] = ["#FFC107", "#FF9800", "#F44336", "#3B82F6", "#9C27B0", "#4CAF50"]

    /// Interpolated flame color for a calorie-burn value (clamped to 0...600).
    static func color(forBurn calories: Double) -> Color {
        let cal = min(max(calories, 0), 600)
        for i in 0..<(stops.count - 1) {
            let lo = stops[i]
            let hi = stops[i + 1]
            // Last segment also absorbs the 540...600 overflow (clamps to the final color).
            if cal <= hi || i == stops.count - 2 {
                let t = hi > lo ? (cal - lo) / (hi - lo) : 0
                return lerp(hexes[i], hexes[i + 1], min(max(t, 0), 1))
            }
        }
        return Color(hex: hexes[hexes.count - 1])
    }

    /// Glow intensity 0...1, verbatim from `glowIntensityForBurn`.
    static func glowIntensity(forBurn calories: Double) -> Double {
        min(max(calories, 0), 600) / 600
    }

    private static func lerp(_ a: String, _ b: String, _ t: Double) -> Color {
        let (ar, ag, ab, _) = Color.rgba(fromHex: a)
        let (br, bg, bb, _) = Color.rgba(fromHex: b)
        return Color(
            .sRGB,
            red: ar + (br - ar) * t,
            green: ag + (bg - ag) * t,
            blue: ab + (bb - ab) * t,
            opacity: 1
        )
    }
}
